tutel/impls/moe_layer.py

```diff
@@ -225,13 +227,24 @@ def forward(self, input: Tensor, gate_index=0, capacity_factor=None, top_k=None,
             result_output.l_aux = None
             return self.result_func(result_output) if self.result_func is not None else result_output
 
+        if torch.is_autocast_enabled():
+            # casts inputs to autocast dtype which enables all2all to be done in low precision
+            if input.device.type == 'cuda':
+                dtype = torch.get_autocast_gpu_dtype()
+            elif input.device.type == 'cpu':
+                dtype = torch.get_autocast_cpu_dtype()
+            elif input.device.type == 'xpu':
+                dtype = torch.xpu.get_autocast_xpu_dtype()  # type: ignore[attr-defined]
+            elif input.device.type == 'hpu':
+                dtype = torch.hpu.get_autocast_hpu_dtype()  # type: ignore[attr-defined]
+            else:
+                raise RuntimeError('User specified autocast device_type must be \'cuda\' or \'cpu\'')
+            input = input.to(dtype=dtype)
+
         original_shape, original_dtype  = input.shape, input.dtype
         assert len(original_shape) >= 2, "Input data must be at least 2D tensor: (s)amples, .., (m)odel_dim"
 
         x = input.reshape(-1, original_shape[-reserve_dims:].numel())
-        for p in self.experts.parameters():
-            x = x.to(p.dtype)
-            break
```

Comment:
I am afraid that removing this may introduce incompatibility with previous behavior. Can you warp it with:
```if not torch.is_autocast_enabled():```?
so that default settings without autocast enabled is not affected by this behavior changes.